### PR TITLE
Openstack: Use juju/http client

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following updates the openstack client factory to use the juju/http
client instead of the one offered by the library. We can reduce a lot of
the client boilerplate by just surfacing the only client.

More work is required to lift it out of the provider into a manifold,
but that work can be done at a later PR.

## See

 - https://github.com/juju/juju/pull/13023
 - https://github.com/juju/juju/pull/13034

### Tests should pass.

```sh
go test -v ./provider/openstack/...
go test -v ./provider/rackspace/...
```

### Bootstrap to an openstack

I used microstack. 
See: https://microstack.run/docs/using-juju

```sh
$ OS_SERIES=bionic
$ IMAGE_ID=$(curl \
   http://cloud-images.ubuntu.com/$OS_SERIES/current/$OS_SERIES-server-cloudimg-amd64.img | \
   microstack.openstack image create \
   --public --container-format=bare --disk-format=qcow2 \
   -f value -c id $OS_SERIES)
$ mkdir ~/simplestreams
$ OS_REGION=microstack
$ KEYSTONE_IP=10.20.20.1
$ juju metadata generate-image -d ~/simplestreams \
   -i $IMAGE_ID -s $OS_SERIES -r $OS_REGION -u http://$KEYSTONE_IP:5000/v3
$ cat <<EOF > microstack.yaml
clouds:
    microstack:
      type: openstack
      auth-types: [access-key,userpass]
      regions:
        microstack:
           endpoint: http://10.20.20.1:5000/v3
EOF
$ juju add-cloud --client microstack -f microstack.yaml
$ source /var/snap/microstack/common/etc/microstack.rc
$ juju autoload-credentials
$ juju bootstrap --bootstrap-series=$OS_SERIES \
  --metadata-source=~/simplestreams \
  --model-default network=test \
  --model-default external-network=external \
  --model-default use-floating-ip=true \
  microstack microstack
```

## Bug reference

 - https://bugs.launchpad.net/juju/+bug/1928182
 - https://bugs.launchpad.net/juju/+bug/1899793

### CODE REVIEW

File: provider/openstack/client.go [hunk 1]
```diff
@@ -172,21 +168,20 @@ func (c *ClientFactory) getClientByAuthMode(authMode identity.AuthMode, cred ide
...
+		Logger: logger,
 	}
+
+	httpClient := jujuhttp.NewClient(
+		jujuhttp.WithSkipHostnameVerification(!sslHostnameVerification),
+		jujuhttp.WithCACertificates(certs...),
+		jujuhttp.WithLogger(logger.Child("http")),
+	)
+	options = append(options, client.WithHTTPClient(httpClient.Client()))
+	return client.NewClient(&cred, authMode, gooseLogger, options...), nil
```
Comment (SimonRichardson) [inline provider/openstack/client.go:186]: This is the meat of it.

File: provider/openstack/client.go [hunk 2]
```diff
@@ -150,11 +150,7 @@ func (c *ClientFactory) getClientState(options ...client.Option) (client.Authent
 
 // getClientByAuthMode creates a new client for the given AuthMode.
 func (c *ClientFactory) getClientByAuthMode(authMode identity.AuthMode, cred identity.Credentials, options ...client.Option) (client.AuthenticatingClient, error) {
-	gooseLogger := gooselogging.LoggoLogger{
-		Logger: loggo.GetLogger("goose"),
-	}
-
-	newClient, err := c.clientFunc(cred, authMode, gooseLogger, c.sslHostnameConfig.SSLHostnameVerification(), c.spec.CACertificates, options...)
+	newClient, err := c.clientFunc(cred, authMode, c.sslHostnameConfig.SSLHostnameVerification(), c.spec.CACertificates, options...)
```
Comment (achilleasa) [inline provider/openstack/client.go:153]: Would it be possible to convert the entire arg list into options?

Comment (SimonRichardson) [reply provider/openstack/client.go:153]: Yeah maybe.

File: provider/openstack/client.go [hunk 3]
```diff
@@ -149,8 +209,13 @@ func (c *ClientFactory) getClientState(options ...client.Option) (client.Authent
...
 
 // getClientByAuthMode creates a new client for the given AuthMode.
-func (c *ClientFactory) getClientByAuthMode(authMode identity.AuthMode, cred identity.Credentials, options ...client.Option) (client.AuthenticatingClient, error) {
-	newClient, err := c.clientFunc(cred, authMode, c.sslHostnameConfig.SSLHostnameVerification(), c.spec.CACertificates, options...)
+func (c *ClientFactory) getClientByAuthMode(authMode identity.AuthMode, cred identity.Credentials, options ...ClientOption) (client.AuthenticatingClient, error) {
+	newClient, err := c.clientFunc(cred, authMode,
+		append(options,
+			WithSkipHostnameVerification(!c.sslHostnameConfig.SSLHostnameVerification()),
+			WithCACertificates(c.spec.CACertificates...),
+		)...,
```
Comment (achilleasa) [inline provider/openstack/client.go:217]: Given that the original slice will be quite small, it is quite possible that you are mutating the caller's value here. I would suggest creating a copy as follows:

```go
append([]ClientOption{
                        WithSkipHostnameVerification(!c.sslHostnameConfig.SSLHostnameVerification()),
                        WithCACertificates(c.spec.CACertificates...),
                }, options...)...,
```

**Review by achilleasa — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**